spirv_msl.cpp

```diff
@@ -10869,6 +10870,57 @@ string CompilerMSL::to_function_args(const TextureFunctionArguments &args, bool
 		{
 			farg_str += ", " + to_expression(lod);
 		}
+		else if (msl_options.sample_dref_lod_array_as_grad && args.dref && imgtype.image.arrayed)
+		{
+			if (msl_options.is_macos() && !msl_options.supports_msl_version(2, 3))
+				SPIRV_CROSS_THROW("Using non-constant 0.0 gradient() qualifier for sample_compare. This is not "
+				                  "supported on macOS prior to MSL 2.3.");
+			// Some Metal devices have a bug where the LoD is erroneously biased upward
+			// when using a level() argument. Since this doesn't happen with gradient2d(),
+			// if we perform the LoD calculation in reverse, we can pass a gradient
+			// instead.
+			// lod = log2(rhoMax/eta) -> exp2(lod) = rhoMax/eta
+			// If we make all of the scale factors the same, eta will be 1 and
+			// exp2(lod) = rho.
+			// rhoX = dP/dx * extent; rhoY = dP/dy * extent
+			// Therefore, dP/dx = dP/dy = exp2(lod)/extent.
+			// (Subtracting 0.5 before exponentiation gives better results.)
```

Comment:
Subtracting 0.5 here seems a bit arbitrary and hacky. That would imply a -0.5 lod bias. Is this required to avoid rounding shenanigans when landing on the mid-point between two levels, e.g. 1.5 or 2.5?

Reply:
No. It's actually required to make the LoD come out right. The CTS test actually fails without this. And I agree--that 0.5 fudge factor is kludgy.

Reply:
Ok, since this is just a hacky workaround, I guess we can go with it ...